File: app/src/main/java/com/baidu/tv/player/repository/PlaylistRepository.java
package com.baidu.tv.player.repository;

import android.content.Context;
import android.util.Log;

import androidx.lifecycle.LiveData;

import com.baidu.tv.player.auth.BaiduAuthService;
import com.baidu.tv.player.database.AppDatabase;
import com.baidu.tv.player.database.PlaylistDao;
import com.baidu.tv.player.database.PlaylistItemDao;
import com.baidu.tv.player.model.Playlist;
import com.baidu.tv.player.model.PlaylistItem;
import com.baidu.tv.player.model.FileInfo;

import java.util.List;
import java.util.ArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.json.JSONArray;
import org.json.JSONException;

/**
 * 播放列表数据仓库
 */
public class PlaylistRepository {
    private static final String TAG = "PlaylistRepository";
    
    private final Context context;
    private final PlaylistDao playlistDao;
    private final PlaylistItemDao playlistItemDao;
    private final ExecutorService executorService;
    
    public PlaylistRepository(Context context) {
        this.context = context.getApplicationContext();
        AppDatabase database = AppDatabase.getInstance(context);
        playlistDao = database.playlistDao();
        playlistItemDao = database.playlistItemDao();
        executorService = Executors.newFixedThreadPool(4);
    }
    
    /**
     * 获取所有播放列表
     */
    public LiveData<List<Playlist>> getAllPlaylists() {
        return playlistDao.getAllPlaylists();
    }
    
    /**
     * 同步获取所有播放列表
     */
    public List<Playlist> getAllPlaylistsSync() {
        return playlistDao.getAllPlaylistsSync();
    }
    
    /**
     * 根据ID获取播放列表
     */
    public LiveData<Playlist> getPlaylistById(long id) {
        return playlistDao.getPlaylistById(id);
    }
    
    /**
     * 同步根据ID获取播放列表
     */
    public Playlist getPlaylistByIdSync(long id) {
        return playlistDao.getPlaylistByIdSync(id);
    }
    
    /**
     * 插入播放列表
     */
    public interface InsertCallback {
        void onSuccess(long id);
        void onError(Exception e);
    }

    public void insertPlaylist(Playlist playlist, InsertCallback callback) {
        executorService.execute(() -> {
            try {
                long id = playlistDao.insert(playlist);
                Log.d(TAG, "播放列表插入成功, ID: " + id);
                if (callback != null) {
                    // 在主线程回调
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(() -> callback.onSuccess(id));
                }
            } catch (Exception e) {
                Log.e(TAG, "播放列表插入失败", e);
                if (callback != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(() -> callback.onError(e));
                }
            }
        });
    }
    
    /**
     * 更新播放列表
     */
    public void updatePlaylist(Playlist playlist, Runnable onSuccess, Runnable onError) {
        executorService.execute(() -> {
            try {
                playlistDao.update(playlist);
                Log.d(TAG, "播放列表更新成功, ID: " + playlist.getId());
                if (onSuccess != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onSuccess);
                }
            } catch (Exception e) {
                Log.e(TAG, "播放列表更新失败", e);
                if (onError != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                }
            }
        });
    }
    
    /**
     * 删除播放列表
     */
    public void deletePlaylist(Playlist playlist, Runnable onSuccess, Runnable onError) {
        executorService.execute(() -> {
            try {
                playlistDao.delete(playlist);
                Log.d(TAG, "播放列表删除成功, ID: " + playlist.getId());
                if (onSuccess != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onSuccess);
                }
            } catch (Exception e) {
                Log.e(TAG, "播放列表删除失败", e);
                if (onError != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                }
            }
        });
    }
    
    /**
     * 获取播放列表的所有项
     */
    public LiveData<List<PlaylistItem>> getPlaylistItems(long playlistId) {
        return playlistItemDao.getItemsByPlaylistId(playlistId);
    }
    
    /**
     * 同步获取播放列表的所有项
     */
    public List<PlaylistItem> getPlaylistItemsSync(long playlistId) {
        return playlistItemDao.getItemsByPlaylistIdSync(playlistId);
    }
    
    /**
     * 插入播放列表项
     */
    public void insertPlaylistItems(List<PlaylistItem> items, Runnable onSuccess, Runnable onError) {
        executorService.execute(() -> {
            try {
                playlistItemDao.insertAll(items);
                Log.d(TAG, "播放列表项插入成功, 数量: " + items.size());
                if (onSuccess != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onSuccess);
                }
            } catch (Exception e) {
                Log.e(TAG, "播放列表项插入失败", e);
                if (onError != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                }
            }
        });
    }
    
    /**
     * 删除播放列表的所有项
     */
    public void deletePlaylistItems(long playlistId, Runnable onSuccess, Runnable onError) {
        executorService.execute(() -> {
            try {
                playlistItemDao.deleteByPlaylistId(playlistId);
                Log.d(TAG, "播放列表项删除成功, playlistId: " + playlistId);
                if (onSuccess != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onSuccess);
                }
            } catch (Exception e) {
                Log.e(TAG, "播放列表项删除失败", e);
                if (onError != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                }
            }
        });
    }
    
    /**
     * 获取播放列表项数量
     */
    public int getPlaylistItemCount(long playlistId) {
        return playlistItemDao.getItemCount(playlistId);
    }

    /**
     * 刷新播放列表
     * @param playlist 要刷新的播放列表
     * @param onSuccess 成功回调
     * @param onError 失败回调
     */
    public void refreshPlaylist(Playlist playlist, Runnable onSuccess, Runnable onError) {
        executorService.execute(() -> {
            try {
                Log.d(TAG, "开始刷新播放列表: " + playlist.getName());
                
                // 获取访问令牌
                String accessToken = BaiduAuthService.getInstance(context).getAccessToken();
                if (accessToken == null || accessToken.isEmpty()) {
                    Log.e(TAG, "未获取到访问令牌，请先登录");
                    if (onError != null) {
                        new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                    }
                    return;
                }
                
                // 1. 获取源目录路径
                String sourcePathsJson = playlist.getSourcePaths();
                if (sourcePathsJson == null || sourcePathsJson.isEmpty()) {
                    Log.w(TAG, "播放列表没有源目录信息，无法刷新");
                    if (onError != null) {
                        new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                    }
                    return;
                }
                
                // 解析源目录路径
                List<String> sourcePaths = new java.util.ArrayList<>();
                try {
                    org.json.JSONArray jsonArray = new org.json.JSONArray(sourcePathsJson);
                    for (int i = 0; i < jsonArray.length(); i++) {
                        sourcePaths.add(jsonArray.getString(i));
                    }
                } catch (org.json.JSONException e) {
                    Log.e(TAG, "解析源目录路径失败", e);
                    if (onError != null) {
                        new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                    }
                    return;
                }
                
                if (sourcePaths.isEmpty()) {
                    Log.w(TAG, "播放列表源目录为空，无法刷新");
                    if (onError != null) {
                        new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                    }
                    return;
                }
                
                // 2. 递归获取所有文件
                List<com.baidu.tv.player.model.FileInfo> allFiles = new java.util.ArrayList<>();
                final java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(sourcePaths.size());
                final java.util.concurrent.atomic.AtomicBoolean hasError = new java.util.concurrent.atomic.AtomicBoolean(false);
                
                for (String path : sourcePaths) {
                    FileRepository.getInstance().fetchFilesRecursive(accessToken, path, new FileRepository.FileListCallback() {
                        @Override
                        public void onSuccess(List<com.baidu.tv.player.model.FileInfo> files) {
                            synchronized (allFiles) {
                                allFiles.addAll(files);
                            }
                            latch.countDown();
                        }

                        @Override
                        public void onFailure(String error) {
                            Log.e(TAG, "获取目录文件失败: " + path + ", error: " + error);
                            // 即使某个目录失败，也尝试继续处理其他目录
                            latch.countDown();
                        }
                    });
                }
                
                try {
                    // 等待所有文件获取完成，最长等待5分钟
                    latch.await(5, java.util.concurrent.TimeUnit.MINUTES);
                } catch (InterruptedException e) {
                    Log.e(TAG, "等待文件获取被中断", e);
                    hasError.set(true);
                }
                
                if (hasError.get()) {
                    if (onError != null) {
                        new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                    }
                    return;
                }
                
                Log.d(TAG, "获取到新文件列表，总数: " + allFiles.size());
                
                // 3. 过滤并转换为播放列表项
                List<PlaylistItem> newItems = new java.util.ArrayList<>();
                long totalDuration = 0;
                int videoCount = 0;
                int imageCount = 0;
                
                // 检查播放列表类型
                int targetMediaType = playlist.getMediaType();
                
                for (int i = 0; i < allFiles.size(); i++) {
                    com.baidu.tv.player.model.FileInfo fileInfo = allFiles.get(i);
                    
                    // 根据播放列表类型过滤
                    boolean shouldAdd = false;
                    if (targetMediaType == 0) { // 混合
                        shouldAdd = fileInfo.isVideo() || fileInfo.isImage();
                    } else if (targetMediaType == 1) { // 视频
                        shouldAdd = fileInfo.isVideo();
                    } else if (targetMediaType == 2) { // 图片
                        shouldAdd = fileInfo.isImage();
                    }
                    
                    if (shouldAdd) {
                        PlaylistItem item = new PlaylistItem();
                        item.setPlaylistId(playlist.getId());
                        item.setFsId(fileInfo.getFsId());
                        item.setFilePath(fileInfo.getPath());
                        item.setFileName(fileInfo.getServerFilename());
                        item.setFileSize(fileInfo.getSize());
                        item.setSortOrder(i);
                        
                        if (fileInfo.isVideo()) {
                            item.setMediaType(1);
                            // 注意：FileInfo中可能没有时长信息，这里暂时设为0
                            // 实际应用中可能需要额外获取视频详情
                            item.setDuration(0);
                            videoCount++;
                        } else if (fileInfo.isImage()) {
                            item.setMediaType(2);
                            imageCount++;
                        }
                        
                        newItems.add(item);
                    }
                }
                
                // 4. 更新数据库
                // 使用事务操作：删除旧项，插入新项，更新播放列表信息
                AppDatabase.getInstance(null).runInTransaction(() -> {
                    // 删除旧项
                    playlistItemDao.deleteByPlaylistId(playlist.getId());
                    
                    // 插入新项
                    playlistItemDao.insertAll(newItems);
                    
                    // 更新播放列表统计信息
                    playlist.setTotalItems(newItems.size());
                    // 如果全是图片，时长为0；如果有视频，这里仅累加已知时长
                    // 由于列表接口可能不返回时长，这里暂不更新时长，或者保持原有时长逻辑
                    
                    // 更新封面：如果原封面是其中的文件，保留；否则使用第一个文件的缩略图
                    if (!newItems.isEmpty()) {
                        String firstItemThumb = newItems.get(0).getFilePath(); // 这里应该使用缩略图地址，但暂时用路径代替
                        // 实际上应该保持原封面，或者提供选项更新封面
                        // 这里简单处理：如果当前没有封面，设置第一个项目的封面
                        if (playlist.getCoverImagePath() == null || playlist.getCoverImagePath().isEmpty()) {
                            // 注意：这里只是路径，实际显示时需要获取缩略图链接
                            // 我们可以暂时不更新封面，或者需要额外的逻辑来获取缩略图
                        }
                    }
                    
                    playlistDao.update(playlist);
                });
                
                Log.d(TAG, "播放列表刷新完成，新文件数: " + newItems.size());
                
                if (onSuccess != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onSuccess);
                }
                
            } catch (Exception e) {
                Log.e(TAG, "刷新播放列表失败", e);
                if (onError != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(onError);
                }
            }
        });
    }
}

File: app/src/main/java/com/baidu/tv/player/kt/repository/PlaylistRepository.kt
package com.baidu.tv.player.kt.repository

import android.util.Log
import com.baidu.tv.player.kt.database.PlaylistDao
import com.baidu.tv.player.kt.database.PlaylistItemDao
import com.baidu.tv.player.kt.model.FileInfo
import com.baidu.tv.player.kt.model.Playlist
import com.baidu.tv.player.kt.model.PlaylistItem
import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import javax.inject.Inject
import javax.inject.Singleton


private const val TAG = "PlaylistRepository"

/**
 * 事务执行器抽象：默认实现委托给 [AppDatabase.withTransaction]，
 * 便于在单测中注入同步执行版本，避免依赖真实 Room 事务调度器。
 */
fun interface TransactionRunner {
    suspend fun runInTransaction(block: suspend () -> Unit)
}

/**
 * 播放列表数据仓库。
 *
 * 全部为 `suspend` 函数，替代 Java 版的 ExecutorService + Handler + CountDownLatch。
 * - 可观察查询返回 Flow（见 DAO）。
 * - [refreshPlaylist] 用协程 + [Mutex] 保护并发刷新。
 */
@Singleton
class PlaylistRepository @Inject constructor(
    private val playlistDao: PlaylistDao,
    private val playlistItemDao: PlaylistItemDao,
    private val fileRepository: FileRepository,
    private val transactionRunner: TransactionRunner,
) {
    private val refreshMutex = Mutex()

    /** 获取所有播放列表（热流）。 */
    fun getAllPlaylists() = playlistDao.getAllPlaylists()

    /** 同步获取所有播放列表。 */
    suspend fun getAllPlaylistsSync(): List<Playlist> = playlistDao.getAllPlaylistsSync()

    /** 根据 ID 获取播放列表（热流）。 */
    fun getPlaylistById(id: Long) = playlistDao.getPlaylistById(id)

    /** 同步根据 ID 获取播放列表。 */
    suspend fun getPlaylistByIdSync(id: Long): Playlist? = playlistDao.getPlaylistByIdSync(id)

    /** 插入播放列表，返回新 id。 */
    suspend fun insertPlaylist(playlist: Playlist): Long {
        val id = playlistDao.insert(playlist)
        Log.d(TAG, "播放列表插入成功, ID: $id")
        return id
    }

    /** 更新播放列表。 */
    suspend fun updatePlaylist(playlist: Playlist) {
        playlistDao.update(playlist)
        Log.d(TAG, "播放列表更新成功, ID: ${playlist.id}")
    }

    /** 删除播放列表。 */
    suspend fun deletePlaylist(playlist: Playlist) {
        playlistDao.delete(playlist)
        Log.d(TAG, "播放列表删除成功, ID: ${playlist.id}")
    }

    /** 获取播放列表项（热流）。 */
    fun getPlaylistItems(playlistId: Long) = playlistItemDao.getItemsByPlaylistId(playlistId)

    /** 同步获取播放列表项。 */
    suspend fun getPlaylistItemsSync(playlistId: Long): List<PlaylistItem> =
        playlistItemDao.getItemsByPlaylistIdSync(playlistId)

    /** 批量插入播放列表项。 */
    suspend fun insertPlaylistItems(items: List<PlaylistItem>) {
        playlistItemDao.insertAll(items)
        Log.d(TAG, "播放列表项插入成功, 数量: ${items.size}")
    }

    /**
     * 原子创建播放列表及其项目。
     *
     * 先插入 [Playlist] 取得 Room 自增 id，再把外部准备好的 [PlaylistItem] 复制为同一 playlistId
     * 后批量插入，避免 UI 层拆分两次写入导致半成品播放列表。
     */
    suspend fun createPlaylistWithItems(playlist: Playlist, items: List<PlaylistItem>): Long {
        var playlistId = 0L
        transactionRunner.runInTransaction {
            playlistId = playlistDao.insert(playlist)
            val itemsWithPlaylistId = items.map { it.copy(playlistId = playlistId) }
            playlistItemDao.insertAll(itemsWithPlaylistId)
            playlistDao.update(playlist.copy(id = playlistId, totalItems = itemsWithPlaylistId.size))
        }
        Log.d(TAG, "播放列表及项目创建成功, ID: $playlistId, 数量: ${items.size}")
        return playlistId
    }

    /** 删除播放列表的所有项。 */
    suspend fun deletePlaylistItems(playlistId: Long) {
        playlistItemDao.deleteByPlaylistId(playlistId)
        Log.d(TAG, "播放列表项删除成功, playlistId: $playlistId")
    }

    /** 获取播放列表项数量。 */
    suspend fun getPlaylistItemCount(playlistId: Long): Int =
        playlistItemDao.getItemCount(playlistId)

    /**
     * 刷新播放列表：递归重新扫描源目录 → 过滤 → 事务删除旧项并插入新项 → 更新统计。
     *
     * 用 [refreshMutex] 保护，避免同一播放列表并发刷新产生脏数据。
     * 单个源目录失败不影响其他目录（与 Java 版一致）。
     *
     * @return 新的播放列表项数量；无 token / 无源目录 / 失败时抛出 [IllegalStateException]。
     */
    suspend fun refreshPlaylist(playlist: Playlist, accessToken: String): Int = refreshMutex.withLock {
        Log.d(TAG, "开始刷新播放列表: ${playlist.name}")

        if (accessToken.isEmpty()) {
            throw IllegalStateException("未获取到访问令牌，请先登录")
        }

        val sourcePathsJson = playlist.sourcePaths
        if (sourcePathsJson.isNullOrEmpty()) {
            throw IllegalStateException("播放列表没有源目录信息，无法刷新")
        }

        val sourcePaths: List<String> = try {
            Gson().fromJson(sourcePathsJson, object : TypeToken<List<String>>() {}.type) ?: emptyList()
        } catch (e: JsonSyntaxException) {
            Log.e(TAG, "解析源目录路径失败", e)
            throw IllegalStateException("解析源目录路径失败", e)
        }
        if (sourcePaths.isEmpty()) {
            throw IllegalStateException("播放列表源目录为空，无法刷新")
        }

        // 递归获取各源目录文件（单目录失败不影响整体）。
        // refreshMutex 已保证同一播放列表不会并发刷新，此处顺序收集即可。
        val allFiles = mutableListOf<FileInfo>()
        var successfulSources = 0
        val failedSources = mutableListOf<String>()
        for (path in sourcePaths) {
            try {
                allFiles.addAll(fileRepository.fetchFilesRecursive(accessToken, path))
                successfulSources++
            } catch (e: Exception) {
                Log.e(TAG, "获取目录文件失败: $path, error: ${e.message}")
                failedSources += path
            }
        }
        // 刷新采用全量快照语义：任一源目录失败都不提交，避免把失败目录中的旧文件误删。
        if (failedSources.isNotEmpty()) {
            throw IllegalStateException("${failedSources.size}个网盘源目录刷新失败，已保留原播放列表")
        }

        Log.d(TAG, "获取到新文件列表，总数: ${allFiles.size}, 成功目录数: $successfulSources")

        // 过滤并转换为播放列表项
        val targetMediaType = playlist.mediaType
        val newItems = ArrayList<PlaylistItem>()
        // 多个源目录可能重叠，按 fs_id（缺失时按 path）去重，避免同一文件重复进入列表。
        val uniqueFiles = allFiles.distinctBy { file ->
            if (file.fsId > 0L) "id:${file.fsId}" else "path:${file.path.orEmpty()}"
        }
        for (fileInfo in uniqueFiles) {
            val shouldAdd = when (targetMediaType) {
                0 -> fileInfo.isVideo() || fileInfo.isImage() // 混合
                1 -> fileInfo.isVideo()
                2 -> fileInfo.isImage()
                else -> fileInfo.isVideo() || fileInfo.isImage()
            }
            if (!shouldAdd) continue

            val itemMediaType = if (fileInfo.isVideo()) 1 else 2
            newItems += PlaylistItem(
                playlistId = playlist.id,
                fsId = fileInfo.fsId,
                filePath = fileInfo.path,
                fileName = fileInfo.serverFilename,
                mediaType = itemMediaType,
                sortOrder = newItems.size,
                duration = 0, // 列表接口不返回时长，播放时再取
                fileSize = fileInfo.size,
            )
        }

        // 刷新后尽量保持当前播放文件；文件已删除时再把索引约束到新列表范围。
        val oldItems = playlistItemDao.getItemsByPlaylistIdSync(playlist.id)
        val currentFsId = oldItems.getOrNull(playlist.lastPlayedIndex)?.fsId
        val refreshedIndex = currentFsId
            ?.let { fsId -> newItems.indexOfFirst { it.fsId == fsId }.takeIf { it >= 0 } }
            ?: playlist.lastPlayedIndex.coerceIn(0, maxOf(0, newItems.lastIndex))

        // 事务：删旧项 → 插新项 → 更新统计和播放位置。成功扫描到空目录时允许同步为空。
        transactionRunner.runInTransaction {
            playlistItemDao.deleteByPlaylistId(playlist.id)
            playlistItemDao.insertAll(newItems)
            playlist.totalItems = newItems.size
            playlist.lastPlayedIndex = refreshedIndex
            playlistDao.update(playlist)
        }

        Log.d(TAG, "播放列表刷新完成，新文件数: ${newItems.size}")
        newItems.size
    }
}
